src/Form/TideMigrationForm.php

```diff
@@ -0,0 +1,155 @@
+   *
+   * @var \Drupal\migrate\Plugin\MigrationPluginManager
+   */
+  protected $pluginManagerMigration;
+
+  /**
+   * The migration definitions.
+   *
+   * @var array
+   */
+  protected $definitions;
+
+  /**
+   * TideMigrationUiForm constructor.
+   *
+   * @param \Drupal\migrate\Plugin\MigrationPluginManager $plugin_manager_migration
+   *   The migration plugin manager.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The config factory service.
+   */
+  public function __construct(MigrationPluginManager $plugin_manager_migration) {
+    $this->pluginManagerMigration = $plugin_manager_migration;
+    $this->definitions = $this->pluginManagerMigration->getDefinitions();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('plugin.manager.migration')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'tide_migration_ui_form';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $options = [];
+    foreach ($this->definitions as $definition) {
+      $migrationInstance = $this->pluginManagerMigration->createStubMigration($definition);
+      if ($migrationInstance->getSourcePlugin() instanceof CSV
+        || $migrationInstance->getSourcePlugin() instanceof Json
+        || $migrationInstance->getSourcePlugin() instanceof Xml
+        || $migrationInstance->getSourcePlugin() instanceof Url) {
+        $id = $definition['id'];
+        $options[$id] = $this->t('%id', [
+          '%id' => $definition['label'] ?? $id,
+        ]);
+      }
+    }
+    $form['tide_migration'] = [
+      '#type' => 'details',
+      '#open' => TRUE,
+      '#title' => $this->t('Tide migration'),
+    ];
+
+    $form['tide_migration']['migrations'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Tide migration group'),
+      '#options' => $options,
+    ];
+    $form['tide_migration']['source_file'] = [
+      '#type' => 'file',
+      '#title' => $this->t('Data source file'),
+      '#description' => t('Select the data file you want to migrate, allowed extensions: csv, json or xml.'),
+    ];
+    $form['tide_migration']['update_existing_records'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Update existing records'),
+      '#default_value' => 1,
+    ];
+    $form['tide_migration']['import'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Migrate'),
+    ];
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    parent::validateForm($form, $form_state);
+    //Support the data files.
+    $validators = ['file_validate_extensions' => ['csv json xml']];
+    // Save file to private file system to protect data file.
+    $file_destination = 'private://';
```

Comment:
@henrytranvan 
I think it doesn't work if we don't have `private` directory

Reply:
Hi @vincent-gao 
The private file system is required for security saving imported data. If the private file system doesn't exist, the error will be shown.

<img width="1677" alt="Screen Shot 2022-08-03 at 2 35 35 pm" src="https://user-images.githubusercontent.com/6699227/182773812-e79d7b76-c4f3-4b2e-807d-b30dd0b1241a.png">

Thanks